### PR TITLE
core: Add support for nested inferrence in IRDL

### PR DESCRIPTION
**Pull request by math-fehr**
This PR allows us to infer the types of some operands and results from the types of other operands and results.

It does so by adding three functions in `AttrConstraint`:
* `get_resolved_variables`: This function returns the set of `ConstraintVar` that will be set during verification.
* `can_infer`: Returns `True` if we can infer a unique attribute represented by the `AttrConstraint` from the assignment of `ConstraintVar`.
* `infer`: Returns the only attribute that satisfy this constraint, given the constraint variables.

We use these methods to get all `ConstraintVar`  from the parsed operand/result types present in the syntax, then infer the constraints on the types that are not in the syntax.

### CODE REVIEW

**Review by PapyChacal — APPROVED** [review]
Yaaay :rocket:

File: xdsl/irdl/declarative_assembly_format.py [hunk 1]
```diff
@@ -136,44 +136,60 @@ def parse(
...
+    ):
+        """
+        Assign constraint variables with values got from the
+        parsed operand and result types.
+        """
+        if any(type is None for type in (*state.operand_types, *state.result_types)):
+            try:
+                for (_, operand_def), operand_type in zip(
+                    op_def.operands, state.operand_types
+                ):
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:150]: Should this be strict?

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:150]: And below

File: xdsl/irdl/irdl.py [hunk 2]
```diff
@@ -268,6 +333,13 @@ def verify(self, attr: Attribute, constraint_vars: dict[str, Attribute]) -> None
         for idx, param_constr in enumerate(self.param_constrs):
             param_constr.verify(attr.parameters[idx], constraint_vars)
 
+    def get_resolved_variables(self) -> set[str]:
+        if len(self.param_constrs) == 0:
+            return set()
```
Comment (superlopuh) [inline xdsl/irdl/irdl.py:338]:
```suggestion
        if not self.param_constrs:
            return set()
```

File: xdsl/irdl/irdl.py [hunk 3]
```diff
@@ -268,6 +333,13 @@ def verify(self, attr: Attribute, constraint_vars: dict[str, Attribute]) -> None
...
+        return set[str].union(
+            *[constr.get_resolved_variables() for constr in self.param_constrs]
+        )
```
Comment (superlopuh) [inline xdsl/irdl/irdl.py:341]: Does this work?
```suggestion
        return {var for for constr in self.param_constrs for var in constr.get_resolved_variables()}
```

File: xdsl/irdl/declarative_assembly_format_parser.py [hunk 4]
```diff
@@ -114,95 +114,62 @@ def parse_format(self) -> FormatProgram:
...
         """
-        for i, (operand, operand_type, (operand_name, _)) in enumerate(
-            zip(self.seen_operands, self.seen_operand_types, self.op_def.operands)
-        ):
-            if not operand:
+        for (
+            seen_operand,
+            seen_operand_type,
+            (operand_name, operand_def),
+        ) in zip(self.seen_operands, self.seen_operand_types, self.op_def.operands):
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format_parser.py:145]: nit: strict here also?

File: xdsl/irdl/irdl.py [hunk 5]
```diff
@@ -209,6 +251,13 @@ def verify(self, attr: Attribute, constraint_vars: dict[str, Attribute]) -> None
                 pass
         raise VerifyException(f"Unexpected attribute {attr}")
 
+    def get_resolved_variables(self) -> set[str]:
+        if len(self.attr_constrs) == 0:
+            return set()
+        return set[str].intersection(
+            *[constr.get_resolved_variables() for constr in self.attr_constrs]
```
Comment (superlopuh) [inline xdsl/irdl/irdl.py:258]: nit:
```suggestion
            *(constr.get_resolved_variables() for constr in self.attr_constrs)
```

**Review by superlopuh — APPROVED** [review]

File: xdsl/irdl/irdl.py (hunk 5 above)
Comment (PapyChacal) [reply xdsl/irdl/irdl.py:258]: Good catch thanks :)

**Pull request merged**